client/blb/client.go

```diff
@@ -1127,109 +1126,123 @@ func (cli *Client) readOneTractReplicated(
 	thisB []byte,
 	thisOffset int64) {
 
+	// TODO(eric): remove this var when we redo bad ts version reporting
 	var badVersionHost string
 
 	reqID := core.GenRequestID()
 	order := rand.Perm(len(tract.Hosts))
+	badReadHosts := make(map[string]bool)
 
 	err := core.ErrAllocHost // default error if none present
+	readFunc := func(host string, n int, thisBLoc []byte) tractResult {
+		// TODO(eric): fill in otherHosts when ts-ts cancellation is done.
+		read, err := cli.tractservers.ReadInto(ctx, host, nil, reqID, tract.Tract, tract.Version, thisBLoc, thisOffset)
+		// TODO(eric): redo this check when we fix error reporting.
+		if err == core.ErrVersionMismatch {
+			badVersionHost = host
+		}
+		if err != core.NoError && err != core.ErrEOF {
+			// TODO(eric): redo bad TS reporting mechanism.
+			return tractResult{0, 0, err, badVersionHost}
+		}
+		log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
+		// In case of short read, i.e., read < len(thisB), we should
+		// pad the result with 0's. This doesn't need to be done
+		// explicitly if 'thisB' given to us starts to be empty.
+		// However, just in case it's not, we will ensure that by
+		// overwritting untouched bytes with 0's.
+		for i := read; i < len(thisBLoc); i++ {
+			thisBLoc[i] = 0
+		}
+		return tractResult{len(thisBLoc), read, err, badVersionHost}
+	}
+
 	if cli.backupReadState.BackupReadBehavior.Enabled {
-		ch := make(chan tractResult)
+		ch := make(chan tractResultRepl)
 
 		ctx, cancel := context.WithCancel(ctx)
```

Comment:
Is the intention for readFunc to use this context? I think it should be. But it's going to use the one from the function arguments...

Reply:
right. I moved readFunc in the outer scope of this block so I could share with the sequential read logic. Will move this.